python/pyspark/sql/types.py

```diff
@@ -938,12 +1016,17 @@ def _infer_type(obj):
                 return MapType(_infer_type(key), _infer_type(value), True)
         else:
             return MapType(NullType(), NullType(), True)
-    elif isinstance(obj, (list, array)):
+    elif isinstance(obj, list):
         for v in obj:
             if v is not None:
                 return ArrayType(_infer_type(obj[0]), True)
         else:
             return ArrayType(NullType(), True)
+    elif isinstance(obj, array):
+        if obj.typecode in _array_type_mappings:
+            return ArrayType(_array_type_mappings[obj.typecode](), False)
+        else:
+            raise TypeError("not supported type: array(%s)" % obj.typecode)
```

Comment:
If we fall back to `_infer_type` here, then `array('q')` will be inferred as something like array of Long, and then the user will see an error from `net.razorvine.pickle` saying that bad array typecode `q`, due to `SPARK-21420`, which seems to me more confusing than explicitly fails here.